src/svm/impls/binary/binary.c

```diff
@@ -1597,6 +1601,214 @@ PetscErrorCode SVMTest_Binary(SVM svm)
+
+  PetscCall(VecWhichGreaterThan(x,lb,&is_xgl));     // get index set for x < C or x < Cp, x < Cm
+  PetscCall(VecWhichLessThan(x,ub,&is_xlu));        // get index set for x > 0
+
+  /* Compute PG_max */
+  PetscCall(ISIntersect(is_yp,is_xlu,&is_int));     // get index set for x < C & y = +1
+  PetscCall(VecGetSubVector(g,is_int,&sg));         // get sub gradient sg = g(x < C & y = +1)
+  PetscCall(VecMin(sg,NULL,&max_1));
+  max_1 *= -1.0;                                    // max_1 = max(-sg) = -min(sg) = -min(g(x < C & y = 1))
+  PetscCall(VecRestoreSubVector(g,is_int,&sg));
+  PetscCall(ISDestroy(&is_int));
+
+  PetscCall(ISIntersect(is_ym,is_xgl,&is_int));    // get index set for x > 0 & y == -1
+  PetscCall(VecGetSubVector(g,is_int,&sg));        // get sub gradient sg = g(x > 0 & y == -1)
+  PetscCall(VecMax(sg,NULL,&max_2));               // max_2 = max(sg) = max(g(x > 0 & y = -1))
+  PetscCall(VecRestoreSubVector(g,is_int,&sg));
+  PetscCall(ISDestroy(&is_int));
+
+  *PG_max = PetscMax(max_1,max_2);
+
+  /* Compute PG_min */
+  PetscCall(ISIntersect(is_ym,is_xlu,&is_int));   // get index set x < C & y = -1
+  PetscCall(VecGetSubVector(g,is_int,&sg));       // min_1 = min(sg) = min(g(x < C & y = -1))
+  PetscCall(VecMin(sg,NULL,&min_1));
+  PetscCall(VecRestoreSubVector(g,is_int,&sg));
+  PetscCall(ISDestroy(&is_int));
+
+  PetscCall(ISIntersect(is_yp,is_xgl,&is_int));  // get index set x > 0 & y = +1
+  PetscCall(VecGetSubVector(g,is_int,&sg));
+  PetscCall(VecMax(sg,NULL,&min_2));
+  min_2 *= -1.0;
+  PetscCall(VecRestoreSubVector(g,is_int,&sg));
+  PetscCall(ISDestroy(&is_int));
+
+  *PG_min = PetscMin(min_1,min_2);
+
+  PetscCall(ISDestroy(&is_yp));
+  PetscCall(ISDestroy(&is_ym));
+  PetscCall(ISDestroy(&is_xgl));
+  PetscCall(ISDestroy(&is_xlu));
+  PetscFunctionReturn(0);
+}
+
+#undef __FUNCT__
+#define __FUNCT__ "SVMConvergedMaximalDualViolation_Binary"
+PetscErrorCode SVMConvergedMaximalDualViolation_Binary(QPS qps,KSPConvergedReason *reason)
+{
+  SVM              svm;
+
+  PetscInt         max_it;
+  PetscReal        atol;
+
+  PetscReal        PGmin,PGmax,v=0.;
+  PetscInt         it;
+
+  PetscFunctionBegin;
+  PetscValidHeaderSpecific(qps,QPS_CLASSID,1);
+
+  PetscCall(QPSGetConvergenceContext(qps,(void*)&svm));
+  PetscCall(QPSGetIterationNumber(qps,&it));
+  PetscCall(QPSGetTolerances(qps,NULL,&atol,NULL,&max_it));
+
+  *reason = KSP_CONVERGED_ITERATING;
+  if (it == 0) PetscFunctionReturn(0);
+
+  PetscCall(SVMComputePGminPGmax_Binary_Private(svm,&PGmin,&PGmax));
+  v = PGmax - PGmin;
+
+  if (it > max_it) {
+    *reason = KSP_DIVERGED_ITS;
+    PetscCall(PetscInfo(qps,"QP solver is diverging (iteration count reached the maximum). Current dual violation %14.12e at iteration %" PetscInt_FMT "\n",(double) v,it));
+    PetscFunctionReturn(0);
+  }
+
+  if ((v <= atol) && PetscAbsReal(PGmax) <= atol && PetscAbsReal(PGmin) <= atol) {
```

Comment:
atol should be set in SVM to something reasonable. Currently it's QPS default `atol = 1e-50`.